File: policy-agent/internal/cli/output.go
package cli

import (
	"encoding/json"
	"fmt"
	"strings"

	"policy-agent/pkg/types"
	"gopkg.in/yaml.v3"
)

// OutputFormat represents the output format
type OutputFormat string

const (
	FormatText OutputFormat = "text"
	FormatJSON OutputFormat = "json"
	FormatYAML OutputFormat = "yaml"
)

// Formatter formats validation results
type Formatter struct {
	format OutputFormat
	color  bool
}

// NewFormatter creates a new output formatter
func NewFormatter(format string, enableColor bool) *Formatter {
	return &Formatter{
		format: OutputFormat(format),
		color:  enableColor,
	}
}

// FormatResult formats a validation result
func (f *Formatter) FormatResult(result *types.ValidationResult) (string, error) {
	switch f.format {
	case FormatJSON:
		return f.formatJSON(result)
	case FormatYAML:
		return f.formatYAML(result)
	default:
		return f.formatText(result), nil
	}
}

// formatJSON formats result as JSON
func (f *Formatter) formatJSON(result *types.ValidationResult) (string, error) {
	data, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return "", fmt.Errorf("failed to marshal JSON: %w", err)
	}
	return string(data), nil
}

// formatYAML formats result as YAML
func (f *Formatter) formatYAML(result *types.ValidationResult) (string, error) {
	data, err := yaml.Marshal(result)
	if err != nil {
		return "", fmt.Errorf("failed to marshal YAML: %w", err)
	}
	return string(data), nil
}

// formatText formats result as human-readable text
func (f *Formatter) formatText(result *types.ValidationResult) string {
	var output strings.Builder

	// Header
	output.WriteString("\n")
	output.WriteString(f.separator("=", 70))
	output.WriteString(f.bold(fmt.Sprintf("Domain: %s (%s)\n", result.Domain, result.ResourceType)))
	output.WriteString(fmt.Sprintf("Resource: %s\n", result.Resource))

	// Status with color
	statusText := fmt.Sprintf("Status: %s", result.Status)
	switch result.Status {
	case types.StatusPassed:
		output.WriteString(f.green(statusText) + "\n")
	case types.StatusFailed:
		output.WriteString(f.red(statusText) + "\n")
	case types.StatusWarning:
		output.WriteString(f.yellow(statusText) + "\n")
	default:
		output.WriteString(statusText + "\n")
	}

	if result.Duration > 0 {
		output.WriteString(fmt.Sprintf("Duration: %v\n", result.Duration))
	}
	if result.File != "" {
		output.WriteString(fmt.Sprintf("File: %s\n", result.File))
	}
	output.WriteString(f.separator("=", 70))

	// Summary
	summary := fmt.Sprintf("\n📊 Summary: %d violations, %d warnings, %d passed\n",
		len(result.Violations), len(result.Warnings), len(result.Passed))
	output.WriteString(f.bold(summary))

	// Violations
	if len(result.Violations) > 0 {
		output.WriteString("\n" + f.red("❌ VIOLATIONS:") + "\n")
		for i, v := range result.Violations {
			output.WriteString(f.formatViolation(i+1, v))
		}
	}

	// Warnings
	if len(result.Warnings) > 0 {
		output.WriteString("\n" + f.yellow("⚠️  WARNINGS:") + "\n")
		for i, w := range result.Warnings {
			output.WriteString(f.formatWarning(i+1, w))
		}
	}

	// Passed checks
	if len(result.Passed) > 0 {
		output.WriteString("\n" + f.green("✅ PASSED:") + "\n")
		for _, p := range result.Passed {
			output.WriteString(fmt.Sprintf("   • %s", p.Policy))
			if p.Message != "" {
				output.WriteString(fmt.Sprintf(": %s", p.Message))
			}
			output.WriteString("\n")
		}
	}

	output.WriteString("\n")
	return output.String()
}

// formatViolation formats a single violation
func (f *Formatter) formatViolation(index int, v types.Violation) string {
	var output strings.Builder

	// Header
	header := fmt.Sprintf("\n%d. [%s] %s\n", index, v.Severity, v.Policy)
	output.WriteString(f.bold(header))

	// Message
	output.WriteString(fmt.Sprintf("   → %s\n", v.Message))

	// Field details
	if v.Field != "" {
		output.WriteString(f.dim(fmt.Sprintf("   Field: %s\n", v.Field)))
		if v.CurrentValue != nil {
			output.WriteString(f.dim(fmt.Sprintf("   Current: %v\n", v.CurrentValue)))
		}
		if v.ExpectedValue != nil {
			output.WriteString(f.dim(fmt.Sprintf("   Expected: %v\n", v.ExpectedValue)))
		}
	}

	// Location
	if v.Line > 0 {
		location := fmt.Sprintf("   Location: line %d", v.Line)
		if v.Column > 0 {
			location += fmt.Sprintf(", column %d", v.Column)
		}
		output.WriteString(f.dim(location + "\n"))
	}

	// Remediation
	if v.Remediation != nil {
		if v.Remediation.AutoFixable {
			output.WriteString(f.green("   🔧 Auto-fixable\n"))
		}
		if v.Remediation.Suggestion != "" {
			output.WriteString(fmt.Sprintf("   💡 Suggestion: %s\n", v.Remediation.Suggestion))
		}
		if v.Remediation.Example != "" {
			output.WriteString("   📝 Example:\n")
			for _, line := range strings.Split(v.Remediation.Example, "\n") {
				output.WriteString(f.dim(fmt.Sprintf("      %s\n", line)))
			}
		}
	}

	return output.String()
}

// formatWarning formats a single warning
func (f *Formatter) formatWarning(index int, w types.Warning) string {
	var output strings.Builder

	header := fmt.Sprintf("\n%d. [%s] %s\n", index, w.Severity, w.Policy)
	output.WriteString(f.bold(header))
	output.WriteString(fmt.Sprintf("   → %s\n", w.Message))

	if w.Field != "" {
		output.WriteString(f.dim(fmt.Sprintf("   Field: %s\n", w.Field)))
	}

	return output.String()
}

// Color and formatting helpers

func (f *Formatter) bold(s string) string {
	if !f.color {
		return s
	}
	return "\033[1m" + s + "\033[0m"
}

func (f *Formatter) dim(s string) string {
	if !f.color {
		return s
	}
	return "\033[2m" + s + "\033[0m"
}

func (f *Formatter) red(s string) string {
	if !f.color {
		return s
	}
	return "\033[31m" + s + "\033[0m"
}

func (f *Formatter) green(s string) string {
	if !f.color {
		return s
	}
	return "\033[32m" + s + "\033[0m"
}

func (f *Formatter) yellow(s string) string {
	if !f.color {
		return s
	}
	return "\033[33m" + s + "\033[0m"
}

func (f *Formatter) separator(char string, length int) string {
	return strings.Repeat(char, length) + "\n"
}

// FormatBatchResults formats multiple validation results
func (f *Formatter) FormatBatchResults(results []*types.ValidationResult) (string, error) {
	if f.format == FormatJSON {
		data, err := json.MarshalIndent(results, "", "  ")
		if err != nil {
			return "", fmt.Errorf("failed to marshal JSON: %w", err)
		}
		return string(data), nil
	}

	if f.format == FormatYAML {
		data, err := yaml.Marshal(results)
		if err != nil {
			return "", fmt.Errorf("failed to marshal YAML: %w", err)
		}
		return string(data), nil
	}

	// Text format - show summary first, then individual results
	var output strings.Builder

	// Overall summary
	totalViolations := 0
	totalWarnings := 0
	totalPassed := 0
	failed := 0
	passed := 0

	for _, result := range results {
		totalViolations += len(result.Violations)
		totalWarnings += len(result.Warnings)
		totalPassed += len(result.Passed)

		if result.Status == types.StatusFailed {
			failed++
		} else if result.Status == types.StatusPassed {
			passed++
		}
	}

	output.WriteString("\n" + f.bold(f.separator("=", 70)))
	output.WriteString(f.bold(fmt.Sprintf("BATCH VALIDATION RESULTS (%d files)\n", len(results))))
	output.WriteString(f.bold(f.separator("=", 70)))
	output.WriteString(fmt.Sprintf("\n📊 Overall Summary:\n"))
	output.WriteString(fmt.Sprintf("   Files: %d total\n", len(results)))
	output.WriteString(f.green(fmt.Sprintf("   Passed: %d\n", passed)))
	output.WriteString(f.red(fmt.Sprintf("   Failed: %d\n", failed)))
	output.WriteString(fmt.Sprintf("   Violations: %d\n", totalViolations))
	output.WriteString(fmt.Sprintf("   Warnings: %d\n", totalWarnings))
	output.WriteString(fmt.Sprintf("   Checks passed: %d\n\n", totalPassed))

	// Individual results
	for _, result := range results {
		output.WriteString(f.FormatResult(result))
	}

	return output.String(), nil
}

File: policy-agent-py/policy_agent/cli/output.py
"""Output formatting for CLI."""

import json
from typing import List
from datetime import datetime

import click
import yaml

from policy_agent.types.result import ValidationResult, Severity


class OutputFormatter:
    """Formatter for validation results."""

    def __init__(self, format_type: str = "text", color: bool = True):
        """Initialize formatter.

        Args:
            format_type: Output format (text, json, yaml)
            color: Enable colored output
        """
        self.format_type = format_type
        self.color = color

    def format_result(self, result: ValidationResult) -> str:
        """Format a single validation result.

        Args:
            result: Validation result to format

        Returns:
            Formatted output string
        """
        if self.format_type == "json":
            return self._format_json(result)
        elif self.format_type == "yaml":
            return self._format_yaml(result)
        else:
            return self._format_text(result)

    def format_batch_results(self, results: List[ValidationResult]) -> str:
        """Format multiple validation results.

        Args:
            results: List of validation results

        Returns:
            Formatted output string
        """
        if self.format_type == "json":
            return self._format_json_batch(results)
        elif self.format_type == "yaml":
            return self._format_yaml_batch(results)
        else:
            return self._format_text_batch(results)

    def _format_text(self, result: ValidationResult) -> str:
        """Format result as human-readable text."""
        output = []

        # Header
        output.append("\n" + "=" * 70)
        output.append(self._bold(f"Domain: {result.domain} ({result.resource_type})"))
        output.append(f"Resource: {result.resource}")

        # Status with color
        status_text = f"Status: {result.status}"
        if result.status == "passed":
            output.append(self._green(status_text))
        elif result.status == "failed":
            output.append(self._red(status_text))
        else:
            output.append(self._yellow(status_text))

        if result.duration:
            output.append(f"Duration: {result.duration:.3f}s")
        if result.file:
            output.append(f"File: {result.file}")

        output.append("=" * 70)

        # Summary
        summary = (
            f"\n📊 Summary: {len(result.violations)} violations, "
            f"{len(result.warnings)} warnings, {len(result.passed)} passed"
        )
        output.append(self._bold(summary))

        # Violations
        if result.violations:
            output.append("\n" + self._red("❌ VIOLATIONS:"))
            for i, v in enumerate(result.violations, 1):
                output.append(self._format_violation(i, v))

        # Warnings
        if result.warnings:
            output.append("\n" + self._yellow("⚠️  WARNINGS:"))
            for i, w in enumerate(result.warnings, 1):
                output.append(self._format_warning(i, w))

        # Passed checks
        if result.passed:
            output.append("\n" + self._green("✅ PASSED:"))
            for p in result.passed:
                msg = f"   • {p.policy}"
                if p.message:
                    msg += f": {p.message}"
                output.append(msg)

        output.append("")
        return "\n".join(output)

    def _format_violation(self, index: int, violation) -> str:
        """Format a single violation."""
        lines = []

        # Header
        severity_color = {
            Severity.CRITICAL: self._red,
            Severity.HIGH: self._red,
            Severity.MEDIUM: self._yellow,
            Severity.LOW: self._dim,
        }.get(violation.severity, self._dim)

        header = f"\n{index}. [{violation.severity.value.upper()}] {violation.policy}"
        lines.append(self._bold(severity_color(header)))

        # Message
        lines.append(f"   → {violation.message}")

        # Field details
        if violation.field:
            lines.append(self._dim(f"   Field: {violation.field}"))
            if violation.current_value is not None:
                lines.append(self._dim(f"   Current: {violation.current_value}"))
            if violation.expected_value is not None:
                lines.append(self._dim(f"   Expected: {violation.expected_value}"))

        # Location
        if violation.line and violation.line > 0:
            location = f"   Location: line {violation.line}"
            if violation.column and violation.column > 0:
                location += f", column {violation.column}"
            lines.append(self._dim(location))

        # Remediation
        if violation.remediation:
            if violation.remediation.auto_fixable:
                lines.append(self._green("   🔧 Auto-fixable"))
            if violation.remediation.suggestion:
                lines.append(f"   💡 Suggestion: {violation.remediation.suggestion}")
            if violation.remediation.example:
                lines.append("   📝 Example:")
                for line in violation.remediation.example.split("\n"):
                    lines.append(self._dim(f"      {line}"))

        return "\n".join(lines)

    def _format_warning(self, index: int, warning) -> str:
        """Format a single warning."""
        lines = []

        header = f"\n{index}. [{warning.severity.value.upper()}] {warning.policy}"
        lines.append(self._bold(header))
        lines.append(f"   → {warning.message}")

        if warning.field:
            lines.append(self._dim(f"   Field: {warning.field}"))

        return "\n".join(lines)

    def _format_text_batch(self, results: List[ValidationResult]) -> str:
        """Format batch results as text."""
        output = []

        # Overall summary
        total_violations = sum(len(r.violations) for r in results)
        total_warnings = sum(len(r.warnings) for r in results)
        total_passed = sum(len(r.passed) for r in results)
        failed = sum(1 for r in results if r.status == "failed")
        passed = sum(1 for r in results if r.status == "passed")

        output.append("\n" + self._bold("=" * 70))
        output.append(self._bold(f"BATCH VALIDATION RESULTS ({len(results)} files)"))
        output.append(self._bold("=" * 70))
        output.append(f"\n📊 Overall Summary:")
        output.append(f"   Files: {len(results)} total")
        output.append(self._green(f"   Passed: {passed}"))
        output.append(self._red(f"   Failed: {failed}"))
        output.append(f"   Violations: {total_violations}")
        output.append(f"   Warnings: {total_warnings}")
        output.append(f"   Checks passed: {total_passed}\n")

        # Individual results
        for result in results:
            output.append(self._format_text(result))

        return "\n".join(output)

    def _format_json(self, result: ValidationResult) -> str:
        """Format result as JSON."""
        output = self._result_to_dict(result)
        return json.dumps(output, indent=2, default=str)

    def _format_json_batch(self, results: List[ValidationResult]) -> str:
        """Format batch results as JSON."""
        output = [self._result_to_dict(r) for r in results]
        return json.dumps(output, indent=2, default=str)

    def _format_yaml(self, result: ValidationResult) -> str:
        """Format result as YAML."""
        output = self._result_to_dict(result)
        return yaml.dump(output, default_flow_style=False, sort_keys=False)

    def _format_yaml_batch(self, results: List[ValidationResult]) -> str:
        """Format batch results as YAML."""
        output = [self._result_to_dict(r) for r in results]
        return yaml.dump(output, default_flow_style=False, sort_keys=False)

    def _result_to_dict(self, result: ValidationResult) -> dict:
        """Convert ValidationResult to dictionary."""
        return {
            "status": result.status,
            "domain": result.domain,
            "resource": result.resource,
            "resource_type": result.resource_type,
            "file": result.file,
            "violations": [
                {
                    "policy": v.policy,
                    "severity": v.severity.value,
                    "message": v.message,
                    "field": v.field,
                    "current_value": v.current_value,
                    "expected_value": v.expected_value,
                    "line": v.line,
                    "column": v.column,
                    "remediation": (
                        {
                            "auto_fixable": v.remediation.auto_fixable,
                            "suggestion": v.remediation.suggestion,
                            "example": v.remediation.example,
                        }
                        if v.remediation
                        else None
                    ),
                }
                for v in result.violations
            ],
            "warnings": [
                {
                    "policy": w.policy,
                    "severity": w.severity.value,
                    "message": w.message,
                    "field": w.field,
                }
                for w in result.warnings
            ],
            "passed": [{"policy": p.policy, "message": p.message} for p in result.passed],
            "timestamp": result.timestamp.isoformat() if result.timestamp else None,
            "duration": result.duration,
        }

    # Color helpers
    def _bold(self, text: str) -> str:
        """Make text bold."""
        if not self.color:
            return text
        return click.style(text, bold=True)

    def _dim(self, text: str) -> str:
        """Make text dim."""
        if not self.color:
            return text
        return click.style(text, dim=True)

    def _red(self, text: str) -> str:
        """Make text red."""
        if not self.color:
            return text
        return click.style(text, fg="red")

    def _green(self, text: str) -> str:
        """Make text green."""
        if not self.color:
            return text
        return click.style(text, fg="green")

    def _yellow(self, text: str) -> str:
        """Make text yellow."""
        if not self.color:
            return text
        return click.style(text, fg="yellow")
